pkg/operator/controller/controller.go

```diff
@@ -351,6 +388,164 @@ func (r *reconciler) ensureInternalRouterServiceForIngress(deployment *appsv1.De
 	return nil
 }
 
+// ensureDefaultCertificateForIngress ensures that a default certificate exists
+// for a given ClusterIngress.
+func (r *reconciler) ensureDefaultCertificateForIngress(deployment *appsv1.Deployment, ci *ingressv1alpha1.ClusterIngress) error {
+	// Normally we would use the manifest factory to build the secret.
+	// However, we want to avoid generating certificates if the secret
+	// already exists.
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("router-certs-%s", ci.Name),
+			Namespace: deployment.Namespace,
+		},
+	}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}, secret)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get secret %s/%s: %v", secret.Namespace, secret.Name, err)
+		}
+
+		ca, err := r.ensureRouterCACertificate()
+		if err != nil {
+			return fmt.Errorf("failed to get CA certificate: %v", err)
+		}
+		hostnames := sets.NewString(fmt.Sprintf("*.%s", *ci.Spec.IngressDomain))
+		cert, err := ca.MakeServerCert(hostnames, 0)
```

Comment:
Rather than a single creation of the CA and cert, you should use the library-go certificate rotators. You'll get creation and rotation on both for free.

CA rotator example:
https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/certrotationcontroller/certrotationcontroller.go#L150

Server cert rotator example: https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/certrotationcontroller/certrotationcontroller.go#L175

Reply:
@mrogers950, thanks for the pointer! It looks like certrotationcontroller is designed to handle a single server certificate whereas we want to have one per router.  Does library-go have something we can use to manage multiple server certifcates, or will it only work for the signing certificate and bundle? Anyway, this is out of scope for 4.0, so I'll add a TODO.

Reply:
Yeah, right now there's one client+server+CA set per controller, so you would need one controller per router.